lib/ast/AST.cpp

```diff
@@ -298,6 +298,36 @@ KOREPattern *KOREAxiomDeclaration::getRightHandSide() const {
   assert(false && "could not compute right hand side of axiom");
 }
 
+KOREPattern *KOREAxiomDeclaration::getRequires() const {
+  if (auto top = dynamic_cast<KOREObjectCompositePattern *>(pattern)) {
+    if (top->getConstructor()->getName() == "\\implies" && top->getArguments().size() == 2) {
+      if (auto equals = dynamic_cast<KOREObjectCompositePattern *>(top->getArguments()[0])) {
+        if (equals->getConstructor()->getName() == "\\and" && equals->getArguments().size() == 2) {
+          equals = dynamic_cast<KOREObjectCompositePattern *>(equals->getArguments()[1]);
+          assert(equals);
+        }
+        if (equals->getConstructor()->getName() == "\\equals" && equals->getArguments().size() == 2) {
+          return equals->getArguments()[0];
+        } else if (equals->getConstructor()->getName() == "\\top" && equals->getArguments().empty()) {
+          return nullptr;
+        }
+      }
+    } else if (top->getConstructor()->getName() == "\\and" && top->getArguments().size() == 2) {
+      if (auto equals = dynamic_cast<KOREObjectCompositePattern *>(top->getArguments()[0])) {
+        if (equals->getConstructor()->getName() == "\\equals" && equals->getArguments().size() == 2) {
+          return equals->getArguments()[0];
+        } else if (equals->getConstructor()->getName() == "\\top" && equals->getArguments().empty()) {
+          return nullptr;
+        }
+      }
+    } else if (top->getConstructor()->getName() == "\\equals" && top->getArguments().size() == 2) {
+      return nullptr;
+    }
+  }
+  assert(false && "could not compute side condition of axiom");
```

Comment:
It seems to me that you should return null here instead of assertion failure. Unless it is the case that for any axiom, we shlould never reach here. But even then the correct thing to do would be to make the `pattern` member of `KOREAxiomDeclalration` to be of `KOREObjectCompositePattern` type and avoid the if branch.

Reply:
I don't really see how changing the type of the declaration would help anything here. As it is, I would rather fail fast if we reach an axiom we don't understand than silently assume it doesn't have a side condition.

Reply:
The point is I cannot tell what the API for this method is. When does it return `nullptr` and when does it abort? It seems to me that it returns `nullptr` when there is no side condition and aborts either when the cast fails or when the axiom's constructor isn't one of `\implies`, `\and`, or `\equals`. Are all of this invariants of our system?

If this is the case here, then the assertion is fine (although it's maybe good to change the type as suggested above to make it even clearer - not necesssary for the pr though). But please change the assertion message to something like `"axiom not well-formed"`, since I think when there is no side condition the expectation is to return `nullptr` rather than abort.